File: 26.cpp
/*

struct TreeNode{
    int val;
    TreeNode *left, *right;
    TreeNode(int x){
        val = x;
        left = right = nullptr;
    }
}

*/

// 判断树2是否为树1的子结构 (空树不是任意一个树的子结构)
bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2){
    if(pRoot1 == nullptr || pRoot2 == nullptr)
        return false;
    bool ans = false;
    if(pRoot1 -> val == pRoot2 -> val)
        ans = isSubtree(pRoot1, pRoot2);
    ans = ans || HasSubtree(pRoot1 -> left, pRoot2) ||
          HasSubtree(pRoot1 -> right, pRoot2);
    return ans;
}

// 树1是否包含树2（头节点相同）
bool isSubtree(TreeNode* pRoot1, TreeNode* pRoot2){
    if(pRoot2 == nullptr)
        return true;
    if(pRoot1 == nullptr)
        return false;
    if(pRoot1 -> val != pRoot2 -> val)
        return false;
    return isSubtree(pRoot1 -> left, pRoot2 -> left) &&
           isSubtree(pRoot1 -> right, pRoot2 -> right);
}

File: 26.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;
 
    public TreeNode(int val) {
        this.val = val;
 
    }
 
}
*/
public class Solution {
    // 判断 root2 是否是 root1 的子结构，空树不算
    public boolean HasSubtree(TreeNode root1,TreeNode root2) {
        if (root1 == null || root2 == null)
            return false;
        if (isSon(root1, root2))
            return true;
        else
            return false;
    }
 
    private boolean isSon(TreeNode root1, TreeNode root2) {
        if (root1 == null || root2 == null)
            return false;
        if (judgeSon(root1, root2))
            return true;
        return isSon(root1.left, root2) || isSon(root1.right, root2);
    }
 
    private boolean judgeSon(TreeNode root1, TreeNode root2) {
        if (root2 == null)
            return true;
        if (root1 == null)
            return false;
        if (root1.val == root2.val && judgeSon(root1.left, root2.left)
            && judgeSon(root1.right, root2.right))
            return true;
        return false;
    }
}

